### ISSUE TITLE
Mosn add support for OpenSSL

### ISSUE THREAD

**Issue by william-zk** [mosn/mosn#1034]
### The New Feature

mosn add support for OpenSSL

### Your scenes

1. the version of mosn crypto package is go-1.9, still not support tls1.3
2. openssl have more powerful feature than go tls
3. mosn need to support GM cipher, but most of GM implements are in openssl
4. use openssl can easier get the ability of hardware speedup

### Your advice

mosn could use cgo to implement the ability of supportting openssl

### Environment

linux, darwin(macos)

- MOSN Version
v1.10 and later

**Comment by taomaree** [comment]
golang 1.13 enabled TLS 1.3 default, maybe upgrade golang version can resolve this problem.

**Comment by william-zk** [comment]
> golang 1.13 enabled TLS 1.3 default, maybe upgrade golang version can resolve this problem.

actually we need more powerful control of TLS，such as some speed up of ciphersuites , and we also need China SM cipher support. Thoose features are difficult to implement in golang. So we choose cgo-openssl

**Comment by hobbytp** [comment]
Will it be any perf issue to use cgo to implement the ability of supportting openssl? will it possible to support both golang TLS and openssl?
Thanks,

**Comment by william-zk** [comment]
> Will it be any perf issue to use cgo to implement the ability of supportting openssl? will it possible to support both golang TLS and openssl?
> Thanks,

Yeah... cgo do have some performance problem, but it depends on the situation, cgo call would be a regardless cost, but most algorithm have a better performance in C. In our practice, algorithm like SM4, RSA would get a lot acceleration when use cgo.
In real tls handshake scene, cipher like ECDHE-RSA-AES128-SHA would get a better performance when use cgo, cipher like ECDHE-ECDSA-AES128-SHA only get half performance of native go tls
In mosn scene, sidecar would hold  a continuous link between each other, cgo call happens frequently in symmetric algorithm, cgo call would get a better performance in SM4 related cipher, but worse than AES related cipher.
Actually what we need is to get the powerful ability of openssl, go native tls is too slow to follow up various of new TLS feature.

**Comment by william-zk** [comment]
> Will it be any perf issue to use cgo to implement the ability of supportting openssl? will it possible to support both golang TLS and openssl?
> Thanks,

Sure we would provide a Switch to you for choosing whether use this feature

**Comment by william-zk** [comment]
Moreover，we find that using cgo-openssl to deal with whole tls process would cause serious performance problem, So we choose use cgo-openssl to take over some cipher calculate process only

**Comment by hobbytp** [comment]
@william-zk Thanks for sharing this kind of info. From my view, main concert is TLS1.3 supporting and performance. Currently envoy proxy (using BoringSsl instead of openSSL) still has problem to support TLS 1.3 in client side, please refer https://github.com/envoyproxy/envoy/issues/9300, which is blocking issue for my case.
